### PR TITLE
Add a bib entry file to make it easy to cite Julia.

### PR DESCRIPTION
**Pull request by ViralBShah**
The proposal is to have a `CITATION.bib` file in the top-level directory of a package. The key for the bib entry should be `<PackageName.jl>-<year>-<letter (if multiple citations in same year)>`. If the `Example.jl` package has two papers in the same year, the citation keys should be `Example-2019-a` and `Example-2019-b`.

Once packages start having `CITATION.bib` files, we can have tools that can take a manifest file and gather all the citations into a single bib file for a particular project.

It is worth noting that package authors can submit to the [Journal of Open Source Software](https://joss.theoj.org/) to have something that is archived and a citable reference.

This is based on a discussion in the #academia channel on slack and being put into action starting with Julia itself.

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
Maybe four-space indent like other code in the repo?

**Comment by StefanKarpinski** [comment]
I would also propose that the citation key be the verbatim package name (not lowercased) and then the year. In this case, it doesn't change since the `julia` repo is already lowercase, but it would be `Example2019` instead of `example2019`. I also wonder about keeping `.jl` in the key since that would ensure no collisions between Julia and non-Julia citations.

**Comment by ViralBShah** [comment]
How about we just drop the year? I wasn't sure if the period would be accepted in the key, but it does work. So the key could just be `Example.jl`.

**Comment by ViralBShah** [comment]
In which case, should this one be `Julia.jl` for uniformity?

File: CITATION.bib [hunk 1]
```diff
@@ -0,0 +1,11 @@
+@article{Julia.jl,
+    title={Julia: A fresh approach to numerical computing},
+    author={Bezanson, Jeff and Edelman, Alan and Karpinski, Stefan and Shah, Viral B},
+    journal={SIAM {R}eview},
+    volume={59},
+    number={1},
+    pages={65--98},
+    year={2017},
+    publisher={SIAM},
+    url={https://doi.org/10.1137/141000671}
```
Comment (giordano) [inline CITATION.bib]: Please, add the field `doi = 10.1137/141000671`, it's correctly used by many biblatex styles

File: CITATION.bib [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
-  journal={SIAM {R}eview},
-  volume={59},
-  number={1},
-  pages={65--98},
-  year={2017},
-  publisher={SIAM},
-  url={https://doi.org/10.1137/141000671}
+@article{Julia.jl,
+    title={Julia: A fresh approach to numerical computing},
+    author={Bezanson, Jeff and Edelman, Alan and Karpinski, Stefan and Shah, Viral B},
```
Comment (StefanKarpinski) [inline CITATION.bib]: Is this really how names are supposed to be listed in bibtex entries? It looks so bizarre. It looks like this paper is the work of
* Bezanson
* Jeff and Edelman
* Alan and Karpinski
* Stefan and Shah
* Viral B

It's like our names got put in a blender and fished out.

Comment (giordano) [reply CITATION.bib]: > Is this really how names are supposed to be listed in bibtex entries

Yes, the "and" keyword separates authors, the comma separates the family name(s) from the given names of each author.

**Comment by non-Jedi** [comment]
I'm not sure I understand the utility of generating a bib file for all
dependencies and recursive dependencies of a package (by looking at a
Manifest). When citing a paper, you wouldn't ever really want to
recurse down the tree of papers that paper cites. That said, if there
is utility there, this approach misses the ability to allow packages
to cite non-packages unless we define some other conventions for entry
keys which seems a highly inelegant path to go down.

File: CITATION.bib (hunk 2 above)
Comment (ViralBShah) [reply CITATION.bib]: Yup. That is right, and bibtex knows how to deal with it also. This bibtex entry is from SIAM - not something I wrote.

**Comment by ViralBShah** [comment]
Recursively generating a bib file in my opinion is ok, because the author will still cite only whatever they need to, but it provides everything at one's fingertips. One can also choose to also just go down one level, or do it manually. Other tools can pull in the bibtex data for inclusion in Julia docs and do many other interesting things with it - perhaps by reviving [BibTeX.jl](https://github.com/stevengj/BibTeX.jl).

There are many other cases related to citing non-packages, citing non-Julia software, and for inclusion in non-latex systems.  This is solving a much simpler problem: having a clear and standard way for package authors to communicate how to cite their software.

**Pull request merged**